pallets/rewards/src/mechanism.rs

```diff
@@ -1,18 +1,53 @@
+use cfg_traits::ops::ensure::EnsureAddAssign;
 use frame_support::traits::tokens::Balance;
 use sp_runtime::{traits::Get, ArithmeticError};
 
 pub mod base;
 pub mod base_with_currency_movement;
 
+pub mod deferred;
+
+pub trait DistributionId: Sized {
+	fn next_id(&mut self) -> Result<Self, ArithmeticError>;
+}
+
+impl DistributionId for () {
+	fn next_id(&mut self) -> Result<Self, ArithmeticError> {
+		Ok(())
+	}
+}
+
+macro_rules! distribution_id_impl {
+	($number:ty) => {
+		impl DistributionId for $number {
+			fn next_id(&mut self) -> Result<Self, ArithmeticError> {
+				self.ensure_add_assign(1)?;
+				Ok(*self)
+			}
+		}
+	};
+}
+
+distribution_id_impl!(u8);
```

Comment:
It opens the door to allow different types of `DistributionId`. The `DistributionId` type is set in the runtime, and a user could ensure that no more than 2^16 or 2^32 distributions would happen and save storage if they use `u16` instead of `u64` (because the `DistributionId` is stored later in accounts/groups/currencies).